scipy/optimize/_sda.py

```diff
@@ -0,0 +1,572 @@
+                    self.state.current_energy = e
+                    self.state.current_location = np.copy(x_visit)
+                    self.xmin = np.copy(self.state.current_location)
+
+                # No improvement since long time
+                if self.not_improved_idx >= self.not_improved_max_idx:
+                    if j == 0 or self.state.current_energy < self.emin:
+                        self.emin = self.state.current_energy
+                        self.xmin = np.copy(self.state.current_location)
+        # End of MarkovChain loop
+
+    def local_search(self):
+        # Decision making for performing a local search
+        # based on Markov chain results
+        # If energy has been improved or no improvement since too long,
+        # performing a local search with the best Markov chain location
+        if self.state_improved:
+            # Global energy has improved, let's see if LS improved further
+            e, x = self.ofw.local_search(self.state.xbest)
+            if e < self.state.ebest:
+                self.not_improved_idx = 0
+                self.state.ebest = e
+                self.state.xbest = np.copy(x)
+                self.state.current_energy = e
+                self.state.current_location = np.copy(x)
+                return
+        # Check probability of a need to perform a LS even if no improvment
+        # (Dual annealing principle)
+        do_ls = False
+        if self.K < 90 * len(self.state.current_location):
+            pls = np.exp(self.K * (self.state.ebest - self.state.current_energy
+                                   ) / self.temperature_step)
+            if pls >= self._rs.random_sample():
+                do_ls = True
+        # Global energy not improved, let's see what LS gives
+        # on the best Markov chain location
+        if self.not_improved_idx >= self.not_improved_max_idx:
+            do_ls = True
+        if do_ls:
+            e, x = self.ofw.local_search(self.xmin)
+            self.xmin = np.copy(x)
+            self.emin = e
+            self.not_improved_idx = 0
+            self.not_improved_max_idx = self.state.current_location.size
+            if e < self.state.ebest:
+                self.state.ebest = self.emin
+                self.state.xbest = np.copy(self.xmin)
+                self.state.current_energy = e
+                self.state.current_location = np.copy(x)
+
+
+class ObjectiveFunWrapper(object):
+    """
+    Class used to wrap around the objective function in order to apply local
+    search and default gradient computation.
+    Default local minimizer is L-BFGS-B
+    """
+    def __init__(self, bounds, func, **kwargs):
+        self.func = func
+        self.nb_fun_call = 0
+        self.kwargs = kwargs
+        self.minimizer = minimize
+        self.fun_args = None
+        lu = list(zip(*bounds))
+        self.lower = np.array(lu[0])
+        self.upper = np.array(lu[1])
```

Comment:
I can't tell that self.lower and self.upper are used anywhere

Reply:
Thanks for this. Good catch, they were used in the default gradient implementation but @antonior92 suggested to use the approx_derivative function that was in _numdiff.py, which was a good idea and I did so. It is not used anymore, I will remove it.

Reply:
I will replace this by bounds because it is needed to pass the bounds as arguments for default minimizer settings.